Added support for std::optional for C# language

christophe-calmejane opened the pull request:
std::optional support for C#

christophe-calmejane committed:
Added support for std::optional for C# language

vadz reviewed (changes requested):
I was curious to see how this worked because, as I wrote, I had my own typemaps for an optional-like class, but after a brief look it looks like this is pretty much based on my own typemaps (it could have been nice to mention this...) except that this doesn't use C# nullable types AFAICS, `cstype` for `optional<T>` is just `T` and not `T?` as in my version.

Unsurprisingly, I think this is wrong, and that the natural representation of `optional<T>` is `T?` and not (possibly null) `T` because the former allows working with them more conveniently at C# level.

IOW there is clearly a lot of useful stuff here (my typemaps didn't include director support), but I think this shouldn't be merged as is and the changes to my original typemaps replacing `T` with `T` should be reverted.

christophe-calmejane commented:
> I was curious to see how this worked because, as I wrote, I had my own typemaps for an optional-like class, but after a brief look it looks like this is pretty much based on my own typemaps (it could have been nice to mention this...)

I've seen quite a lot of tentative implementation of optional that all look pretty much the same, not sure if *yours* is shared somewhere but after testing quite a lot of solutions to be as clean as possible, this is what came up entirely rewritten based on an implementation I made quite a few years ago.

> except that this doesn't use C# nullable types AFAICS, `cstype` for `optional<T>` is just `T` and not `T?` as in my version.
> Unsurprisingly, I think this is wrong, and that the natural representation of `optional<T>` is `T?` and not (possibly null) `T` because the former allows working with them more conveniently at C# level.

Indeed, this is a last minute change to acommodate with Mono restriction that doesn't understand this (needs C#>=8), but then I forgot it is not a C# nullable reference anymore. I guess I need to find a way to properly switch from one or other implementation, certainly with a define. But then how to have proper unit testing since we only have one version of the C# compiler?

christophe-calmejane commented on Lib/csharp/std_optional.i:
```diff
@@ -0,0 +1,256 @@
+%enddef
+
+
+#if !defined(SWIG_NO_STD_OPTIONAL_DEFAULT_TYPES)
```
If I include std_optional.i I would expect it to just work by default for basic types, but I'm open to discussion. This is much like including cstdint, you have all types defined directly. Either way it's a decision to be taken (or is there some precedence we can rely on?)

vadz replied:
To me, this is similar to `std_vector.i` predefining `vector<int>` etc — which is actually also done for C# currently but was IMO a mistake and just creates a gratuitous difference between it and the other languages.

vadz commented:
> I've seen quite a lot of tentative implementation of optional that all look pretty much the same, not sure if _yours_ is shared somewhere but after testing quite a lot of solutions to be as clean as possible, this is what came up entirely rewritten based on an implementation I made quite a few years ago.

Well, this contains several comments that I wrote verbatim (which is how I recognized it, the writing style looked familiar), so I have little doubts that this is based on [my typemaps](https://gist.github.com/vadz/c6a9cf0c62d8db7ad687992735a9ae08) originally written in 2013 which I posted [at least here](https://github.com/swig/swig/issues/1307#issuecomment-469258404) and I believe in some other places, so you can easily compare the two versions.

> Indeed, this is a last minute change to acommodate with Mono restriction that doesn't understand this (needs C#>=8)

I don't understand it, as I wrote on the mailing list, we definitely didn't use C# 8 in 2013 and yet my typemaps worked just fine with whatever version we did use back then. Why do you think C# 8 is needed?

christophe-calmejane commented:
> > I've seen quite a lot of tentative implementation of optional that all look pretty much the same, not sure if _yours_ is shared somewhere but after testing quite a lot of solutions to be as clean as possible, this is what came up entirely rewritten based on an implementation I made quite a few years ago.
> 
> Well, this contains several comments that I wrote verbatim (which is how I recognized it, the writing style looked familiar), so I have little doubts that this is based on [my typemaps](https://gist.github.com/vadz/c6a9cf0c62d8db7ad687992735a9ae08) originally written in 2013 which I posted [at least here](https://github.com/swig/swig/issues/1307#issuecomment-469258404) and I believe in some other places, so you can easily compare the two versions.

Indeed you are right. This is a 5y long work that was started by an old colleague, I didn't know he started (or was inspired) from your work. Sorry about that I didn't know when I took over as it wasn't mentioned. I'm curious though if it's been so long since you did that work, why isn't it in the official repository?

christophe-calmejane commented:
> > Indeed, this is a last minute change to acommodate with Mono restriction that doesn't understand this (needs C#>=8)
> 
> I don't understand it, as I wrote on the mailing list, we definitely didn't use C# 8 in 2013 and yet my typemaps worked just fine with whatever version we did use back then. Why do you think C# 8 is needed?

Well, I'm not at all a C# expert so I'm not entirely sure what is wrong. When I try to use nullable reference like you suggest, I get CI errors like this:
```
cpp17_optional/TestOptionals.cs(85,10): error CS0453: The type `cpp17_optionalNamespace.Struct' must be a non-nullable value type in order to use it as type parameter `T' in the generic type or method `System.Nullable<T>'
cpp17_optional/TestOptionals.cs(91,33): error CS0453: The type `cpp17_optionalNamespace.Struct' must be a non-nullable value type in order to use it as type parameter `T' in the generic type or method `System.Nullable<T>'
```

I don't get any of these errors when compiling a C# project (outside the CI) that uses a recent C# compiler (ie. dotnet), and also if I try to force a C#<8 in my project I correctly get this error:
```
Program.cs(71,42): error CS8107: Feature 'nullable reference types' is not available in C# 7.0. Please use language version 8.0 or greater.
```

That's where I assumed Mono is not implementing C#>=8, or at least not nullable reference. Again, not an expert, so it might just be something totally different, but still everything is working fine when I define the typemap as nullable reference for all my projects, just the CI that don't accept it.
I'd be glad to revert to nullable reference as it was the original plan, so if you have any clue what is wrong, feel free to advice.

On a side note, I only have access to macOS for testing, it's a quite a pain to rely on the CI for testing small changes (although for compilation only and not running it might be sufficient. I'll try to setup things again locally for that purpose).

christophe-calmejane commented:
Also, there is one more thing that I cannot seem to fix when using nullable references, I get this warning with my modern C# compiler:
```
warning CS8669: The annotation for nullable reference types should only be used in code within a '#nullable' annotations context. Auto-generated code requires an explicit '#nullable' directive in source.
```

And I don't know how to add the requested `#nullable enable` directive using typemap (adding a csimports typemap has no effect, I don't know if there is a way to add things to the generated .cs files). Ofc this directive makes no sense for a mono compiler (might even be an error), this the reason to be able to target C#<8 or C#>=8

christophe-calmejane committed:
Improved c# optional with generation time options

- SWIG_STD_OPTIONAL_DEFAULT_TYPES to auto-bind arithmetic types and
  string
- SWIG_STD_OPTIONAL_USE_NULLABLE_REFERENCE_TYPES to bind to nullable
  reference types for C#8

christophe-calmejane committed:
Updated unit tests

christophe-calmejane committed:
[c# std_optional] Added more unit tests

christophe-calmejane committed:
[c# std_optional] Fixed memory leaks and unhandled typemap

christophe-calmejane committed:
[c# std_optional] Added more unit tests

christophe-calmejane committed:
[c# std_optional] Fixed incorrect typemap for 'in' lvalue typemap

christophe-calmejane committed:
[c# std_optional] Added possibility to change internal class modifier

christophe-calmejane commented:
@wsfulton @ojwb This PR has one remaining issue in this code https://github.com/swig/swig/blob/8fd4df01bd4111e5f7322fe4f9d4817ffd6f0c6d/Source/Modules/csharp.cxx#L4252-L4257

When using a `nullable reference type` this type of line is generated:
```csharp
private static global::System.Type[] swigMethodTypes1 = new global::System.Type[] { typeof(Rect?) };
```
which fails to compile with this error
```
error CS8639: The typeof operator cannot be used on a nullable reference type
```

I tested by manually changing the generated line to `typeof(Rect)` and it is working just fine and `SwigDerivedClassHasMethod` is happy.

Can you hint me to what needs to be changed in `csharp.cxx` module to properly detect the type is a nullable reference and remove the `?` (but only for this, we still want to have `?` for basic types and structs).
Thanks

ojwb commented on Examples/test-suite/cpp17_optional.i:
```diff
@@ -0,0 +1,334 @@
+        void               setInt(int i) { _numberInt = i; }
+        std::optional<int> getIntOpt() const { return _numberIntOpt; }
+        void               setIntOpt(std::optional<int> i) { _numberIntOpt = i; }
+        // Flaot optional testing
```
Typo: "Flaot" -> "Float"

ojwb commented on Lib/csharp/std_optional.i:
```diff
@@ -0,0 +1,329 @@
+    #include <optional>
+%}
+
+%include <stdint.i>
```
Implicitly including `stdint.i` opens something of a can of worms.  By default SWIG aims to generate code independent of the platform it is run on, which can then be built anywhere.  A few places violate this, and `stdint.i` is one of them as the code generated is tailored to the system it will be built on (the problem is that for most target languages SWIG needs to work in terms of `int`, `long`, `long long`, etc which of those e.g. `int64_t` corresponds to depends on the platform).

So I think it would be better to avoid doing this by default.

christophe-calmejane replied:
Oh dear I didn't get the notification on April 2nd when Olly reviewed the PR. I can try without the inclusion and see if that works. I've been using this code for more than a year in multiple projects now.

ojwb commented:
> Can you hint me to what needs to be changed in `csharp.cxx` module to properly detect the type is a nullable reference and remove the `?` (but only for this, we still want to have `?` for basic types and structs).

I think this is addressed by #3114?

christophe-calmejane committed:
Code review changes

christophe-calmejane commented:
Hi, is there still any blocker for this one?
I'd like to add optional for lua, but it depends on this one, and on directors for lua. It's a bit complicated to handle all theses branches. Thanks

wsfulton commented:
I think I've seen a bunch of other `std::optional` merge requests which ought to be looked at together with this one to ensure we have a common set of features for across all the target languages. I'll get up to speed on this merge request next.

@vadz, any objections to using this merge request over any of your older patches (that this work was derived from)?

vadz commented:
But less optimal simply types are a big problem, our C# API works with them since 15 years and it would be a huge step back to require all its users to use much less convenient code (without even speaking of requiring them to change something working perfectly well at all).

christophe-calmejane committed:
Full rewrite of C# optional for primitive types and enums

christophe-calmejane commented:
I've spent the last 2 days trying to find a solution for satisfy this but it's very tricky as there is always something breaking (either enums, directors, nullable ref, ...). The most difficult part is solving directors handling for enums (I absolutely need it) that #3114 is not solving, but I may have a better solution that #3114 (I worked with the author of that PR and it was made for this one) by using a new attribute (so we only rely on what the typemap declares instead of checking types in swig, which doesn't handle enums).

I hope I can find a solution, otherwise I still think it's better to have support for optionals, even if not optimal for primitive types, than no support at all (we'll have the option to improve it in a next release). The good thing is you said you already have it in use for 15y, so you won't be forced to use this one if it breaks your codebase, as you probably have your own file anyway.

christophe-calmejane committed:
C#: Fix std::optional type conversion for unsigned integer enums

When using std::optional with enum types that have unsigned underlying types
(uint8_t, uint16_t, uint32_t, uint64_t), the generated C# code was passing
unsigned values (byte, ushort, uint, ulong) to Marshal.WriteXXX methods that
expect signed types (sbyte, short, int, long).

This caused compilation errors like:
  error CS1503: Argument #2 cannot convert 'ulong' expression to type 'long'
